runtime/doc/lsp.txt

```diff
@@ -1812,27 +1825,6 @@ stop({bufnr}, {client_id})                    *vim.lsp.semantic_tokens.stop()*
 ==============================================================================
 Lua module: vim.lsp.handlers                                    *lsp-handlers*
 
-hover({_}, {result}, {ctx}, {config})               *vim.lsp.handlers.hover()*
```

Comment:
off-topic: do we have a issue explaining the plan for `vim.lsp.handlers`, is it intended to be fully obsolete? or parts of it will continue to be relevant?

Reply:
Needs a tracking issues. Here's a quick brain dump...

Ideally, I would prefer some path on which it is removed, as the model it is a part of just doesn't work for a lot of LSP methods; so it would be nice if we can replace it with a model that does.

AFAIK the handlers serve two purposes:
1. as extension points for plugins
2. as hook points to augment responses from servers. This is mostly useful to fix bugs/incorrect behaviour from some servers.

For 1, I don't think it's the best model anyway, as it isn't particularly hard to issue your own request.

For 2, I think this is something that would be nice to keep.

As a rough proposal:
- Deprecate vim.lsp.handlers entirely.
- Add `client:on_request(fn)` and `client:on_response(fn)` to the client object which can be used to intercept communication between the client and the server.
  - This does have some overlap with the `LspRequest` autocmd, however that doesn't support modifying the `params`/`result` sent from the client/server.

Reply:
Raised #30982